File: frontend/src/types/reading.ts
export type MetricName = 'temp_analog_c' | 'temp_digital_c' | 'humidity' | 'aqi' | 'tvoc' | 'eco2' | 'pm1' | 'pm25' | 'pm10'

export interface SensorReading {
  id: number
  recorded_at: string
  temp_analog_c: number | null
  temp_digital_c: number | null
  humidity: number | null
  aqi: number | null
  tvoc: number | null
  eco2: number | null
  pm1: number | null
  pm25: number | null
  pm10: number | null
}

export interface HistoryPoint { recorded_at: string; value: number }
export interface ReadingHistoryResponse { metric: MetricName; points: HistoryPoint[] }

export interface ReadingStatsResponse {
  metric: MetricName
  range: number
  count: number
  minimum: number | null
  average: number | null
  maximum: number | null
}

File: backend/app/models/reading.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field


MetricName = Literal[
    "temp_analog_c",
    "temp_digital_c",
    "humidity",
    "aqi",
    "tvoc",
    "eco2",
    "pm1",
    "pm25",
    "pm10",
]


class SensorReading(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    recorded_at: datetime
    temp_analog_c: Optional[float] = None
    temp_digital_c: Optional[float] = None
    humidity: Optional[float] = None
    aqi: Optional[int] = Field(default=None, ge=1, le=5)
    tvoc: Optional[int] = Field(default=None, ge=0)
    eco2: Optional[int] = Field(default=None, ge=0)
    pm1: Optional[float] = Field(default=None, ge=0)
    pm25: Optional[float] = Field(default=None, ge=0)
    pm10: Optional[float] = Field(default=None, ge=0)


class HistoryPoint(BaseModel):
    recorded_at: datetime
    value: Union[float, int]


class ReadingHistory(BaseModel):
    metric: MetricName
    points: list[HistoryPoint]


class ReadingStats(BaseModel):
    metric: MetricName
    range: int = Field(ge=1, le=168)
    count: int = Field(ge=0)
    minimum: Optional[float] = None
    average: Optional[float] = None
    maximum: Optional[float] = None
